javascript/src/applier.ts

```diff
@@ -0,0 +1,402 @@
+ * {@link https://html.spec.whatwg.org/multipage/rendering.html HTML Rendering}.
+ * See also {@link actionForElement}.
+ */
+const domActions: {[name: string]: DomAction} = {
+  // Hidden elements
+  // https://html.spec.whatwg.org/multipage/rendering.html#hidden-elements
+  AREA: DomAction.Skip,
+  BASE: DomAction.Skip,
+  BASEFONT: DomAction.Skip,
+  DATALIST: DomAction.Skip,
+  HEAD: DomAction.Skip,
+  LINK: DomAction.Skip,
+  META: DomAction.Skip,
+  NOEMBED: DomAction.Skip,
+  NOFRAMES: DomAction.Skip,
+  PARAM: DomAction.Skip,
+  RP: DomAction.Skip,
+  SCRIPT: DomAction.Skip,
+  STYLE: DomAction.Skip,
+  TEMPLATE: DomAction.Skip,
+  TITLE: DomAction.Skip,
+  NOSCRIPT: DomAction.Skip,
+
+  // Flow content
+  // https://html.spec.whatwg.org/multipage/rendering.html#flow-content-3
+  HR: DomAction.Break,
+  // Disable if `white-space: pre`.
+  LISTING: DomAction.Skip,
+  PLAINTEXT: DomAction.Skip,
+  PRE: DomAction.Skip,
+  XMP: DomAction.Skip,
+
+  // Phrasing content
+  // https://html.spec.whatwg.org/multipage/rendering.html#phrasing-content-3
+  BR: DomAction.Break,
+  RT: DomAction.Skip,
+
+  // Form controls
+  // https://html.spec.whatwg.org/multipage/rendering.html#form-controls
+  INPUT: DomAction.Skip,
+  SELECT: DomAction.Skip,
+  BUTTON: DomAction.Skip,
+  TEXTAREA: DomAction.Skip,
+
+  // Other elements where the phrase-based line breaking should be disabled.
+  // https://github.com/google/budoux/blob/main/budoux/skip_nodes.json
+  ABBR: DomAction.Skip,
+  CODE: DomAction.Skip,
+  IFRAME: DomAction.Skip,
+  TIME: DomAction.Skip,
+  VAR: DomAction.Skip,
+};
+
+/**
+ * Determine the action for an element.
+ * @param element An element to determine the action for.
+ * @returns The {@link domActions} for the element.
+ */
+function actionForElement(element: Element): DomAction {
+  const action = domActions[element.nodeName];
+  if (action !== undefined) return action;
+
+  // jsdom does not have `getComputedStyle`.
+  if (typeof getComputedStyle === 'function') {
+    const style = getComputedStyle(element);
+    switch (style.whiteSpace) {
+      case 'nowrap':
+      case 'pre':
+        return DomAction.Skip;
+    }
+
+    const display = style.display;
+    assert(display);
+    if (display !== 'inline') return DomAction.Block;
+  }
+  return DomAction.Inline;
+}
+
+/**
+ * Represents a "paragraph", broken by block boundaries or forced breaks.
+ *
+ * A CSS
+ * {@link https://drafts.csswg.org/css2/#inline-formatting inline formatting context}
+ * is usually a "paragraph", but it can be broken into multiple paragraphs by
+ * forced breaks such as `<br>`.
+ */
+class Paragraph {
+  element: HTMLElement;
+  textNodes: Text[] = [];
+
+  constructor(element: HTMLElement) {
+    this.element = element;
+  }
+
+  hasText(): boolean {
+    return this.textNodes.length > 0;
+  }
+}
+
+/**
+ * Applies the BudouX to the given DOM.
+ *
+ * This class has following advantages over
+ * {@link Applier.applyToElement}.
```

Comment:
Probably this should be `parser.applyElement` instead?

Reply:
Fixed, thanks!